pkg/go/graph/graph.go

```diff
@@ -39,4 +41,47 @@ func (g *AuthorizationModelGraph) GetDOT() string {
 	return string(dotRepresentation)
 }
 
-// TODO add graph traversals, cycle detection, etc.
+// CycleInformation encapsulates whether the graph has cycles.
+type CycleInformation struct {
+	hasCyclesAtCompileTime bool
+	hasCyclesAtRuntime     bool
+}
+
+func (g *AuthorizationModelGraph) nodeListHasNonComputedEdge(nodeList []graph.Node) bool {
+	for i, nodeI := range nodeList {
+		for _, nodeJ := range nodeList[i+1:] {
+			otherEdges := g.Lines(nodeI.ID(), nodeJ.ID())
```

Comment:
I'm curious why you have a double for loop? Can we not do this like this?

```
for i := 0; i < len(nodeList) - 1; i++ {
	this := nodeList[i]
	next := nodeList[i+1]
        // ... 
}

Reply:
It does not cover all potential paths - the node list only tells you what nodes are involved in the loop.  
Assume nodes 1,2,3,4 are returned.
There can be paths between Node1 and Node3 (or node 2 and node 4).  We will need to examine whether these paths are computed or not too.

Reply:
Ohhh OK!